File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        queue=[start]
        visited=set()
        while(queue):
            curr=queue.pop(0)
            if curr in visited:
                continue
            if arr[curr]==0:
                return True
            visited.add(curr)
            if curr+arr[curr]<len(arr):
                queue.append(curr+arr[curr])
            if curr-arr[curr]>=0:
                queue.append(curr-arr[curr])
        return False
            
            
        
        

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        queue<int>Q;
        unordered_set<int>visited;
        Q.push(start);
        while(!Q.empty())
        {
        int curr=Q.front();Q.pop();
            
        if(arr[curr]==0)return true;
            if(curr-arr[curr] >=0 and curr-arr[curr]<arr.size() and visited.find(curr-arr[curr])==visited.end())
            {
                Q.push(curr-arr[curr]);
                visited.insert(curr-arr[curr]);
            }
            if(curr+arr[curr]>=0 and curr+arr[curr]<arr.size() and visited.find(curr+arr[curr])==visited.end())
            {
                Q.push(curr+arr[curr]);
                visited.insert(curr+arr[curr]);
            }
            
        }
        return false;
    }
};